internal/lsp/foldingrange.go

```diff
@@ -0,0 +1,176 @@
+package lsp
+
+import (
+	"strings"
+
+	"github.com/open-policy-agent/opa/ast"
+
+	"github.com/styrainc/regal/internal/lsp/opa/scanner"
+	"github.com/styrainc/regal/internal/lsp/opa/tokens"
+	"github.com/styrainc/regal/internal/lsp/types"
+)
+
+type stack []scanner.Position
+
+func (s stack) Push(p scanner.Position) stack {
+	return append(s, p)
+}
+
+func (s stack) Pop() (stack, scanner.Position) {
+	l := len(s)
+
+	return s[:l-1], s[l-1]
+}
+
+func TokenFoldingRanges(policy string) []types.FoldingRange {
+	scn, err := scanner.New(strings.NewReader(policy))
+	if err != nil {
+		panic(err)
+	}
+
+	var lastPosition scanner.Position
+
+	curlyBraceStack := stack{}
+	bracketStack := stack{}
+	parensStack := stack{}
+
+	foldingRanges := make([]types.FoldingRange, 0)
+
+	for {
+		token, position, _, errors := scn.Scan()
+
```

Comment:
[style] With all these,

```go
if something {
    // ...
    continue
}
if somethingElse {
    // ...
    continue
}
```

we could also use a conditionless switch
```go
for {
  token, position, _, errors := scn.Scan()
  switch {
    case token == tokens.EOF || len(errors) > 0:
        break
    case token= tokens.LBRACE:
        curlyBraceStack = curlyBraceStack.Push(position)
    case token == tokens.RBrace && len(curlyBraceStack) > 0:
        // ...
  }
}
```

Reply:
Ah, yeah, that's better! I'll make it so, thanks 👍

Reply:
It didn't work with the `break` condition as it won't break out of the loop but the switch. So I moved that on top and kept the rest.

Reply:
ah ok 😅 could have used labels, but this is fine, too!